microsoft-edge/progressive-web-apps-chromium/offline.md

```diff
@@ -0,0 +1,183 @@
+
+> [!IMPORTANT]
+> Web Storage is a synchronous process and, as such, is not available for use within worker threads such as Service Worker.  
+
+Heavy usage may create performance issues for your application.  
+
+Web Storage comes in two flavors: `localStorage` and `sessionStorage`.  Each is maintained as a separate data store isolated to the domain that created it.  As you may expect, `sessionStorage` persists only for the duration of the browsing session \(for example, while the browser is open, inclusive of refresh and restore\); whereas `localStorage` persists until it is removed by the developer, the user, or the browser \(when under storage pressure\).  The API is identical for both storage types.  
+
+```javascript
+var data = {
+    title: document.querySelector("[property='og:title']").getAttribute("content"),
+    description: document.querySelector( "meta[name='description']" ).getAttribute("content")
+};
+localStorage.setItem( window.location, JSON.stringify(data) );
+```  
+
+This basic code sample grab metadata about the current page and store it in a JavaScript object.  Then it stores that value as JSON in `localStorage` \(using `setItem()` method\), assigned to a key equal to the current `window.location` URL.  You may retrieve the information from `localStorage` by using `getItem()` method.  
+
+An excerpt from an offline page that enumerates cached pages and extracts that metadata to build a rich link list.  
+
+```javascript
+caches.open( "pages" )
+      .then( cache => {
+    cache.keys()
+         .then( keys => {
+        if ( keys.length )
+        {
+            keys.forEach( insertOfflineLink );
+        }
+    })
+});
+
+function insertOfflineLink( request ) {
+    var data = JSON.parse( localStorage.getItem( request.url ) );
+    // if data exists and it is not the offline page
+    if ( data && request.url.indexOf('offline') < 0  )
+    {
+        // build a link and insert it into the page
+    }
+}
+```  
+
+Here, the `insertOfflineLink()` method passes the URL of the request into the `localStorage.getItem()` method to retrieve any stored metadata that may be stored when the page is viewed.  If that data is found, the method builds it and inserts the requisite markup to display it.  
+
+## Network Information  
+
+In addition to being able to store information for offline use, you may find it quite helpful to know when a network connection is available in order to synchronize data or even inform each of your users that the network status has changed.  There are a few options for getting the information you need in this regard.  
+
+### navigator.onLine  
+
+The `navigator.onLine` property is a simple boolean that lets you know the current status of the network.  If the value is set to `true`, the user is online; if the value is set to `false`, the user is offline.  
```

Comment:
```suggestion
The `navigator.onLine` property is a boolean that lets you know the current status of the network. If the value is `true`, the user is online, otherwise the user is offline.
```